File: Leetcode/Array/findPeakElement-162.ts
/**
 * https://leetcode.com/problems/find-peak-element/
 * Đi tìm vị trí index trong mảng sao cho tại vị trí index thì đó là phần tử đỉnh
 * Phần tử đỉnh là phần tử lớn hơn số bên trái và bên phải của nó. Giống với đỉnh núi/đỉnh trong đồ thị
 * Vẽ biểu đồ ra là hiểu.
 */
function findPeakElement(nums: number[]): number {
	const bs = (nums: Array<number>, left: number, right: number): number => {
		const middle: number = Math.floor((left + right) / 2)
		if (nums[middle] < nums[middle - 1]) return bs(nums, left, middle)
		else if (nums[middle] < nums[middle + 1]) return bs(nums, middle, right)
		else return middle
	}
	const left: number = 0,
		right: number = nums.length
	return bs(nums, left, right)
}
console.log('index: ', findPeakElement([1]))

File: WithC/findPeakElement-162.cpp
#include <iostream>
#include <vector>

int findPeakElement(std::vector<int> &nums);
int binarySearch(std::vector<int> &nums, int left, int right);

int main()
{
    std::vector<int> nums = {1};
    int output = findPeakElement(nums);
    std::cout << output << std::endl;
    return 0;
}
int findPeakElement(std::vector<int> &nums)
{
    return binarySearch(nums, 0, nums.size());
};
int binarySearch(std::vector<int> &nums, int left, int right)
{
    int middle = (left + right) / 2;
    if (nums[middle] < nums[middle - 1])
        return binarySearch(nums, left, middle);
    else if (nums[middle] < nums[middle + 1])
        return binarySearch(nums, middle, right);
    else
        return middle;
}